pilot/pkg/model/push_context.go

```diff
@@ -1903,16 +1911,68 @@ func (ps *PushContext) initEnvoyFilters(env *Environment) {
 		return in < jn
 	})
 
-	ps.envoyFiltersByNamespace = make(map[string][]*EnvoyFilterWrapper)
 	for _, envoyFilterConfig := range envoyFilterConfigs {
 		efw := convertToEnvoyFilterWrapper(&envoyFilterConfig)
-		if _, exists := ps.envoyFiltersByNamespace[envoyFilterConfig.Namespace]; !exists {
-			ps.envoyFiltersByNamespace[envoyFilterConfig.Namespace] = make([]*EnvoyFilterWrapper, 0)
-		}
 		ps.envoyFiltersByNamespace[envoyFilterConfig.Namespace] = append(ps.envoyFiltersByNamespace[envoyFilterConfig.Namespace], efw)
 	}
 }
 
+// pre computes envoy filters per namespace
+func (ps *PushContext) updateEnvoyFilters(env *Environment, changedEnvoyFilters []ConfigKey, previousIndex map[string][]*EnvoyFilterWrapper) {
+	ps.envoyFiltersByNamespace = maps.Clone(previousIndex)
+
+	for _, changedEnvoyFilter := range changedEnvoyFilters {
+		envoyFilterConfig := env.Get(gvk.EnvoyFilter, changedEnvoyFilter.Name, changedEnvoyFilter.Namespace)
+		// copy existing envoy filters since we're going to modify the slice
+		envoyFilters := slices.Clone(ps.envoyFiltersByNamespace[changedEnvoyFilter.Namespace])
+
+		// was deleted
+		if envoyFilterConfig == nil {
+			envoyFilters := slices.FilterInPlace(envoyFilters, func(envoyFilter *EnvoyFilterWrapper) bool {
+				return envoyFilter.Name != changedEnvoyFilter.Name
+			})
+			if len(envoyFilters) == 0 {
+				// the slice is empty, better to just remove the key from the map
+				delete(ps.envoyFiltersByNamespace, changedEnvoyFilter.Namespace)
+			} else {
+				ps.envoyFiltersByNamespace[changedEnvoyFilter.Namespace] = envoyFilters
+			}
+
+			continue
+		}
+
+		envoyFilter := slices.FindFunc(envoyFilters, func(envoyFilter *EnvoyFilterWrapper) bool {
+			return envoyFilter.Name == changedEnvoyFilter.Name
+		})
+
+		if envoyFilter != nil {
+			*envoyFilter = convertToEnvoyFilterWrapper(envoyFilterConfig)
+		} else {
+			envoyFilters = append(envoyFilters, convertToEnvoyFilterWrapper(envoyFilterConfig))
+			sort.Slice(envoyFilters, func(i, j int) bool {
+				ifilter := envoyFilters[i]
+				jfilter := envoyFilters[j]
+				if ifilter.Priority != jfilter.Priority {
+					return ifilter.Priority < jfilter.Priority
+				}
+
```

Comment:
i cannt reuse the same function because they take different objects, I could parametrize all the needed fields, but that would mean a method taking like 8 or 10 parameters, didn't seem too good.